### PR TITLE
Add timestamp instrumentation for latency measurement

### PR DESCRIPTION
**Pull request by oteffahi**
## Description
This PR adds the ability to record timestamps on messages along the Zenoh network path, notably to measure latency and identify bottlenecks in Zenoh routing.

### What does this PR do?
Adds `TimestampStack` extension to application messages (Push, Query, Response) and expose API to configure instrumentation and access collected timestamps.

Sister PR for Python bindings: https://github.com/eclipse-zenoh/zenoh-python/pull/739

---
## 🏷️ Label-Based Checklist

Based on the labels applied to this PR, please complete these additional requirements:

**Labels:** `new feature`

## 🆕 New Feature Requirements

Since this PR adds a new feature:

- [x] **Feature scope documented** - Clear description of what the feature does and why it's needed
- [x] **Minimum necessary code** - Implementation is as simple as possible, doesn't overcomplicate the system
- [x] **New APIs well-designed** - Public APIs are intuitive, consistent with existing APIs
- [x] **Comprehensive tests** - All functionality is tested (happy path + edge cases + error cases)
- [x] **Examples provided** - Usage examples in code comments or separate example files
- [x] **Documentation added** - New docs explaining the feature, its use cases, and API
- [x] **Feature flag considered** - Can the feature be enabled/disabled for gradual rollout?
- [x] **Performance impact assessed** - Memory, CPU, storage implications measured
- [x] **Integration tested** - Feature works with existing features

**Consider:** Can this feature be split into smaller, incremental PRs?

**Instructions:**
1. Check off items as you complete them (change `- [ ]` to `- [x]`)
2. The PR checklist CI will verify these are completed

*This checklist updates automatically when labels change, but preserves your checked boxes.*

### CODE REVIEW

File: zenoh/src/api/timestamp_stack.rs [hunk 1]
```diff
@@ -0,0 +1,543 @@
...
+/// The struct is `#[non_exhaustive]` to allow adding new fields in the future.
+#[non_exhaustive]
+#[zenoh_macros::unstable]
+pub struct TsStackContext {
+    /// The Zenoh ID of the current node.
+    pub zid: ZenohId,
+    /// The mode of the current node (router, peer, or client).
+    pub whatami: WhatAmI,
+    /// The interception point identifier (e.g., `zenoh_protocol::network::timestamp_stack::interception_point::SEND`).
+    pub interception_point: InterceptionPoint,
```
Comment (OlivierHecart) [inline zenoh/src/api/timestamp_stack.rs]: The presence of this mandatory field (as well as this structure name) make the TimestampCallback very specific to sample interception timestamping. It prevents us to reuse the callback in other contexts. 
Maybe we should rename this struct and remove this field to make it more generic ?

File: zenoh/src/api/timestamp_stack.rs [hunk 2]
```diff
@@ -0,0 +1,543 @@
...
+}
+
+/// Type alias for the user-defined timestamp callback.
+///
+/// The callback receives a [`TsStackContext`] and returns the raw timestamp bytes
+/// to be pushed onto the timestamp stack.
+#[zenoh_macros::unstable]
+pub type GetTimestampCallback = Arc<dyn Fn(TsStackContext) -> Vec<u8> + Send + Sync>;
```
Comment (OlivierHecart) [inline zenoh/src/api/timestamp_stack.rs]: Ideally, the fact that the Fn is stored in an Arc should be hidden.

**Review by OlivierHecart — APPROVED** [review]
